### PR TITLE
Add checking for tool distribution in Tier 1

### LINKED ISSUE #54483: Verify that the tools are actually distributed on beta/stable channels
In #54206 it was noted that RLS was missing in `x86_64-pc-windows-gnu`, but toolstate didn't see the issue because it only checks `x86_64-pc-windows-msvc` and `x86_64-unknown-linux-gnu`. We ought to ensure that the tools are not forgotten at least on all tier-1 platforms.

The simplest fix right now is to add a check that if a tool failed to build, the whole dist job should be aborted. Currently we *allowed* tool dist to fail because of this:

https://github.com/rust-lang/rust/blob/e7b5ba8661aa844a06c37f22d7af0afb1807d347/src/bootstrap/dist.rs#L1166-L1169

similar for Clippy:

https://github.com/rust-lang/rust/blob/e7b5ba8661aa844a06c37f22d7af0afb1807d347/src/bootstrap/dist.rs#L1245-L1252

and rustfmt:

https://github.com/rust-lang/rust/blob/e7b5ba8661aa844a06c37f22d7af0afb1807d347/src/bootstrap/dist.rs#L1323-L1331

What we should do is modify the `or_else` branch to `panic!` unless we are configured to allow omitting these tools (i.e. in the nightly channel). (Consider refactor this into the same function to avoid writing the same thing 5 times.)

While it is possible to check the release channel directly inside rustbuild, I consider this a bad coupling. It is better to add a configuration variable `--enable-missing-tools`, and modify `src/ci/run.sh` to pass this into `$RUST_CONFIGURE_ARGS` when 

1. `$RUST_RELEASE_CHANNEL = "nightly"`, or
2. the env var `$DIST_REQUIRE_ALL_TOOLS` is undefined

You may see #39824 for how to add a config to rustbuild.

After this issue is fixed, we'll ensure RLS/Clippy/Rustfmt will be present on all tier-1 platforms on beta/stable. In the future we should record and alert the tool developers when they fail to build on nightly or lower-tier platforms, but let's get the immediate issue fixed first...

### PR DESCRIPTION
**Pull request by pvdrz**
This fixes https://github.com/rust-lang/rust/issues/54483
r? @kennytm

### CODE REVIEW

**Comment by kennytm** [comment]
Seems legit 

```
[00:03:42] /checkout/src/ci/run.sh: line 79: [: too many arguments
```

File: config.toml.example [hunk 1]
```diff
@@ -481,3 +481,7 @@
 # as the one built on Windows will contain backslashes in paths causing problems
 # on linux
 #src-tarball = true
+#
+
+# Wether to allow failures when building tools
```
Comment (kennytm) [inline config.toml.example]: s/Wether/Whether/

File: src/bootstrap/config.rs [hunk 2]
```diff
@@ -375,6 +377,7 @@ impl Config {
         config.rust_codegen_backends = vec![INTERNER.intern_str("llvm")];
         config.rust_codegen_backends_dir = "codegen-backends".to_owned();
         config.deny_warnings = true;
+        config.missing_tools = false;
```
Comment (kennytm) [inline src/bootstrap/config.rs]: `config.missing_tools` is never read from `toml.dist.missing_tools`. 

For reference, please search for "src_tarball" in this file to see how the value is propagated from `toml.dist.*` into `config.*`.

Comment (christianpoveda) [reply src/bootstrap/config.rs]: Sure I will :)

Comment (christianpoveda) [reply src/bootstrap/config.rs]: I did `set(&mut config.missing_tools, t.missing_tools` when `Some(t) = toml.dist` just like `src_tarball`

**Comment by kennytm** [comment]
@bors retry 

3 hour timeout

**Pull request merged**